wgpu-core/src/command/ray_tracing.rs

```diff
@@ -0,0 +1,1223 @@
+#[cfg(feature = "trace")]
+use crate::device::trace;
+#[cfg(feature = "trace")]
+use crate::ray_tracing::{
+    TlasInstance, TraceBlasBuildEntry, TraceBlasGeometries, TraceBlasTriangleGeometry,
+    TraceTlasInstance, TraceTlasPackage,
+};
+use crate::{
+    device::queue::TempResource,
+    global::Global,
+    id::CommandEncoderId,
+    init_tracker::MemoryInitKind,
+    lock::RwLockReadGuard,
+    ray_tracing::{
+        tlas_instance_into_bytes, BlasAction, BlasBuildEntry, BlasGeometries,
+        BuildAccelerationStructureError, TlasAction, TlasBuildEntry, TlasPackage,
+        ValidateBlasActionsError, ValidateTlasActionsError,
+    },
+    resource::{Blas, Tlas},
+    FastHashSet,
+};
+
+use wgt::{math::align_to, BufferAddress, BufferUsages, Features};
+
+use super::{BakedCommands, CommandBufferMutable};
+use crate::ray_tracing::BlasTriangleGeometry;
+use crate::resource::{AccelerationStructure, Buffer, Fallible, Labeled, StagingBuffer, Trackable};
+use crate::scratch::ScratchBuffer;
+use crate::snatch::SnatchGuard;
+use crate::storage::Storage;
+use crate::track::PendingTransition;
+use hal::BufferUses;
+use std::ops::Deref;
+use std::sync::Arc;
+use std::{cmp::max, num::NonZeroU64, ops::Range};
+
+type BufferStorage<'a> = Vec<(
+    Arc<Buffer>,
+    Option<PendingTransition<BufferUses>>,
+    Option<(Arc<Buffer>, Option<PendingTransition<BufferUses>>)>,
+    Option<(Arc<Buffer>, Option<PendingTransition<BufferUses>>)>,
+    BlasTriangleGeometry<'a>,
+    Option<Arc<Blas>>,
+)>;
+
+type BlasStorage<'a> = Vec<(
+    Arc<Blas>,
+    hal::AccelerationStructureEntries<'a, dyn hal::DynBuffer>,
+    u64,
+)>;
+
+// TODO: Get this from the device (e.g. VkPhysicalDeviceAccelerationStructurePropertiesKHR.minAccelerationStructureScratchOffsetAlignment) this is currently the largest possible some devices have 0, 64, 128 (lower limits) so this could create excess allocation (Note: dx12 has 256).
+const SCRATCH_BUFFER_ALIGNMENT: u32 = 256;
+
+impl Global {
+    pub fn command_encoder_build_acceleration_structures_unsafe_tlas<'a>(
```

Comment:
These functions are _very_ similar, is there any way we can define them in terms of each other?

Reply:
Defining one in terms of the other could be a lot of work and I'm not entirely sure of the use-case of `command_encoder_build_acceleration_structures_unsafe_tlas` vs adding `into_hal` methods. I think that if we *need* to define one in terms of the other it might be more worthwhile to just add `into_hal` methods instead, but perhaps @JMS55 could comment on the use-case for this function?

Reply:
Alright, could you just add a note as to why they're currently separate, and we can re-evaluate later.

Reply:
No idea either, I didn't use the unsafe functions ever.

However the general problem wgpu's as_hal() functions have is that you have no way of hooking into wgpu's synchronization code, so there's no real way to use the hal objects anyways.

Reply:
I don't think there is much synchronization code in build_acceleration_structures, but there is no `as_hal` method currently so this should be reconsidered in a later PR. I will add a todo talking about reconsidering this once `as_hal` for blases and tlases is in.

Reply:
It's more like if you build the acceleration structures, you don't want the build commands to be reordered after the draw commands that use them.

Can be worked-around with multiple command encoders though probably, and you really want to build BLAS's on an async compute queue anyways, although that's tricky with wgpu vulkan due to vulkan not letting you request additional queues past device initialization.